### PR TITLE
SAK-41172 Group lock refactor

### PR DESCRIPTION
**Pull request by bgarciaentornos**
-Added a group lock service and ddbb table.
-Updated previous references.
-New methods (UNTESTED): references that are locked by a group / groups
that are locked by a reference.

### CODE REVIEW

File: kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/GroupLockServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	@Transactional(readOnly = true)
+	public List<String> getGroupsLockedByItem(String itemId) {
+		if(StringUtils.isEmpty(itemId)){
+			return null;
+		}
+        Query query = sessionFactory.getCurrentSession().getNamedQuery("findGroupsLockedByItem");
+        query.setString("itemId", itemId);
+        return (List<String>) query.list();
+	}
+	private String getGroupLockProperty(LockMode lockMode){
```
Comment (ern) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/GroupLockServiceImpl.java]: this method can be removed

File: kernel/kernel-impl/src/main/webapp/WEB-INF/grouplock-components.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    <!-- class="org.sakaiproject.messagebundle.impl.CachingMessageBundleServiceImpl"-->
+	<bean id="org.sakaiproject.site.api.GroupLockService"
+          class="org.springframework.transaction.interceptor.TransactionProxyFactoryBean">
+        <property name="target">
+            <bean class="org.sakaiproject.site.impl.GroupLockServiceImpl">
+                <property name="sessionFactory" ref="org.sakaiproject.springframework.orm.hibernate.GlobalSessionFactory"/>
+            </bean>
+        </property>
+        <property name="transactionManager" ref="org.sakaiproject.springframework.orm.hibernate.GlobalTransactionManager"/>
```
Comment (ern) [inline kernel/kernel-impl/src/main/webapp/WEB-INF/grouplock-components.xml]: this isn't need if your using tx:annotation-driven and marking the service with methods with @Transactional

File: kernel/api/src/main/java/org/sakaiproject/site/api/GroupLock.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+@Table(name = "SAKAI_GROUP_LOCK")
+@NamedQueries({
+	@NamedQuery(name = "findGroupWithLock", query = "from GroupLock where groupId = :groupId and lockMode = :groupLockProperty"),
+	@NamedQuery(name = "findItemsLockingGroup", query = "select itemId from GroupLock where groupId = :groupId"),
+	@NamedQuery(name = "findGroupsLockedByItem", query = "select itemId from GroupLock where itemId = :itemId")
+})
+
+@Data
+@NoArgsConstructor
+@IdClass(GroupLock.class)
```
Comment (ern) [inline kernel/api/src/main/java/org/sakaiproject/site/api/GroupLock.java]: Is this a strange use of @IdClass usually its another class not the same

Comment (bgarciaentornos) [reply kernel/api/src/main/java/org/sakaiproject/site/api/GroupLock.java]: If all the values of the class match with the id, it can be used this way.

**Comment by bgarciaentornos** [comment]
> Should the GroupLockService should probably be folded into AuthzGroupService?

I believe it's better to keep it separated, so things are easier to look at for anybody else in the future. The service might grow more and having all in one big service is sometimes confusing... But anyway, whatever you decide.

File: kernel/kernel-impl/src/main/webapp/WEB-INF/grouplock-components.xml (hunk 2 above)
Comment (bgarciaentornos) [reply kernel/kernel-impl/src/main/webapp/WEB-INF/grouplock-components.xml]: I tried removing it, but it was causing errors. So I decided to remove the tx:annotation-driven tag.

**Comment by bgarciaentornos** [comment]
> Codacy Here is an overview of what got changed by this pull request:
> 
> ```diff
> Issues
> ======
> + Solved 1
> - Added 3
>            
> 
> Complexity decreasing per file
> ==============================
> + site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/rsf/GroupListProducer.java  -1
> + site-manage/site-manage-group-section-role-helper/tool/src/java/org/sakaiproject/site/tool/helper/managegroupsectionrole/rsf/GroupDelProducer.java  -1
>          
> ```
> 
> See the complete overview on [Codacy](https://app.codacy.com/app/sakaiproject/sakai/pullRequest?prid=2877537&bid=10700261)

That review was wrong, I launched the re-analyze and the added issues disappeared.

**Comment by ern** [comment]
closing this one as we'll be taking #6536

**Pull request closed without merging**